### PR TITLE
KRun's Search Functionality

### PR DESCRIPTION
**Pull request by msaxena2**
--search and --pattern support for KRun. WIP, not yet ready for review. Also solves #1667 .

### CODE REVIEW

**Comment by grosu** [comment]
@msaxena2 , please check with @andreistefanescu , because he said during our K meeting that this should be easy.

File: kernel/src/main/java/org/kframework/krun/KRunOptions.java [hunk 1]
```diff
@@ -249,6 +249,7 @@ public SearchType searchType() {
 
     public static final String DEFAULT_PATTERN = "<generatedTop> B:Bag </generatedTop> [anywhere]";
 
+    public static final String DEFAULT_KORE_PATTERN = "<generatedTop> B:Bag </generatedTop> [anywhere]";
```
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/krun/KRunOptions.java]: unfortunately this isn't quite right anymore... You will need to grab the actual name of the actual top cell from the compiled definition and use that. Also the [anywhere] should not be needed anymore.

File: kernel/src/main/java/org/kframework/krun/modes/KRunExecutionMode.java [hunk 2]
```diff
@@ -37,6 +38,17 @@ public KRunExecutionMode(KRunOptions kRunOptions, KExceptionManager kem, FileUti
...
+            if (kRunOptions.pattern != null) {
+                pattern = KRun.compilePattern(files, kem, kRunOptions.pattern, kRunOptions, compiledDefinition, Source.apply("<command line>"));
+            }
+            else {
+                pattern = KRun.compilePattern(files, kem, kRunOptions.DEFAULT_KORE_PATTERN, kRunOptions, compiledDefinition, Source.apply("<command line>"));
+            }
+
+            return new SearchResult(rewriter.search(k, Optional.ofNullable(kRunOptions.depth), Optional.ofNullable(kRunOptions.bound), pattern));
+        }
         if (kRunOptions.exitCodePattern != null) {
```
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/krun/modes/KRunExecutionMode.java:55]: you also need to handle the case where --pattern is specified without --search

**Comment by msaxena2** [comment]
@dwightguth @andreistefanescu @grosu Please review.

File: kernel/src/main/java/org/kframework/krun/KRun.java [hunk 3]
```diff
@@ -151,6 +157,35 @@ public static void prettyPrint(CompiledDefinition compiledDef, OutputModes outpu
         }
     }
 
+    public static void prettyPrintSubstitution(Map<? extends KVariable, ? extends K> subst,
```
Comment (grosu) [inline kernel/src/main/java/org/kframework/krun/KRun.java:211]: @msaxena2 , we have agreed that all public functions will need to have javadoc.  If the name is not obvious, then also add some explanation of what it does.  I thought we made it all clear that all public functions should be documented.

@dwightguth please confirm that you are going to reject code which is not documented well enough.  At minimum, that should include good names and javadoc for public methods.

File: kernel/src/main/java/org/kframework/krun/KRun.java [hunk 4]
```diff
@@ -73,20 +74,25 @@ public int run(CompiledDefinition compiledDef, KRunOptions options, Function<Mod
 
         if (result instanceof K) {
             prettyPrint(compiledDef, options.output, s -> outputFile(s, options), (K) result);
-
-            if (options.exitCodePattern != null) {
```
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/krun/KRun.java:77]: why is this gone?

File: kernel/src/main/java/org/kframework/krun/modes/KRunExecutionMode.java [hunk 5]
```diff
@@ -34,18 +38,29 @@ public KRunExecutionMode(KRunOptions kRunOptions, KExceptionManager kem, FileUti
...
+        if (kRunOptions.search()) {
+            if (pattern == null) {
+                pattern = new Rule(KORE.KVariable("X"), BooleanUtils.TRUE, BooleanUtils.TRUE, KORE.Att());
+                parsedPattern = pattern;
+            }
+            return new SearchResult(rewriter.search(k, Optional.ofNullable(kRunOptions.depth), Optional.ofNullable(kRunOptions.bound), pattern, kRunOptions.searchType()), parsedPattern);
+        }
         if (kRunOptions.exitCodePattern != null) {
-            Rule pattern = KRun.compilePattern(files, kem, kRunOptions.exitCodePattern, kRunOptions, compiledDefinition, Source.apply("<command line: --exit-code>"));
+            pattern = KRun.compilePattern(files, kem, kRunOptions.exitCodePattern, kRunOptions, compiledDefinition, Source.apply("<command line: --exit-code>"));
```
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/krun/modes/KRunExecutionMode.java]: something's not quite right here... if I specify --pattern and --exit-code but not --search, it isn't going to behave the way one would expect.

**Comment by dwightguth** [comment]
Can you resolve conflicts so that we can merge this?

**Comment by msaxena2** [comment]
@dwightguth I've resolved the conflicts. Can you please go through the diff once again to ensure nothing's out of place? Thanks.

**Pull request merged**